confirm.c

```diff
@@ -395,38 +398,102 @@ static struct confirm_query * confirm_pkt_parse4(const struct packet *pkt)/*{{{*
 	return q;
 } /*}}}*/
 
+// TODO (rafael): this is a draft, confirm_pkt_parse6 with TCP support
 static struct confirm_query * confirm_pkt_parse6(const struct packet *pkt)/*{{{*/
 {
 	assert(pkt->ip->ip_v == 6);
-	if(pkt->ipv6->ip_nh != IPPROTO_ICMP6) return NULL;
-	if(pkt->icmpv6->icmp_type != ICMP6_ECHOREPLY &&
-			pkt->icmpv6->icmp_type != ICMP6_TIMXCEED) {
-		return NULL;
-	}
 
 	uint16_t icmpid = 0;
 	uint16_t data = 0;
 	uint8_t traffic_class = 0;
 	uint32_t flow_label = 0;
+	int probe_type = 0; // probe_type can hold 1 (icmp) or 2 (tcp)
 
 	struct sockaddr_in6 dst;
 	dst.sin6_family = AF_INET6;
-	if(pkt->icmpv6->icmp_type == ICMP6_ECHOREPLY) {
+
+	if(pkt->ipv6->ip_nh==IPPROTO_TCP) {
+		// SYN-ACK
+		// TODO (rafael): draft
+		if(pkt->tcp->th_flags != 0x012) return NULL;
+		probe_type = 2;
 		memcpy(&dst.sin6_addr, &pkt->ipv6->ip_src, sizeof(dst.sin6_addr));
-		icmpid = ntohs(pkt->icmpv6->id);
-		data = ntohs(pkt->icmpv6->seq);
-	} else if(pkt->icmpv6->icmp_type == ICMP6_TIMXCEED) {
-		if(pkt->icmpv6->icmp_code != ICMP_TIMXCEED_INTRANS) return NULL;
-		struct libnet_ipv6_hdr *rip;
-		struct libnet_icmpv6_hdr *ricmp;
-		rip = (struct libnet_ipv6_hdr *)(pkt->payload);
-		ricmp = (struct libnet_icmpv6_hdr *)(pkt->payload + LIBNET_IPV6_H);
-		memcpy(&dst.sin6_addr, &rip->ip_dst, sizeof(dst.sin6_addr));
-		uint32_t flags = *(uint32_t *)(rip->ip_flags);
-		traffic_class = (flags & 0x0FF00000) >> 20;
-		flow_label = (flags & 0x000FFFFF);
-		icmpid = ntohs(ricmp->id);
-		data = ntohs(ricmp->seq);
+		uint32_t tcp_sequence_number = ntohl(pkt->tcp->th_ack) - 1;
+		icmpid = (uint16_t) ((tcp_sequence_number >> 16) & 0x0000FFFF);
```

Comment:
this variable has to change name, as this is a TCP probe (and a TCP reply). an idea would be to rename `data` and `icmpid` to denote the information that they store.